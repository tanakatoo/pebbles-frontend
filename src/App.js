import './styles/app.css';
import NavBar from "./components/common/NavBar"
import "./styles/app.css"
import { useSelector } from 'react-redux';
import MyRoutes from "./MyRoutes"

function App() {
  //get the font from the store
  const font = useSelector(state => state.langFont.font)

  //we have to hardcode the string of the font because tailwind doesn't know what to use at build
  return (
    <div className={`App ${font == 'font-poppins' ? 'font-poppins' : 'font-notoJP'}`}>
      <NavBar />
      <MyRoutes />
    </div >
  );
}

export default App;
