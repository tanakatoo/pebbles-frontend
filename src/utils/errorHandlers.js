import { rest } from 'msw'

let loginResponse = {
    token: 'abcdefg',
    user: {
        id: 2,
        username: 'hello',
        role: 'regular',
        premium_join_date: '',
        premium_end_date: '',
        free_trial_start_date: '',
    }
}



export const errorHandlers = [

    //login
    rest.post('http://localhost:3001/auth/login', (req, res, ctx) => {
        const failLogin = {
            "error": {
                "message": "INVALID_CREDENTIALS",
                "status": 401
            }
        }
        console.log(res(ctx.json([failLogin])))
        return res(ctx.json([failLogin]));
    }),

    rest.post('http://localhost:3001/email', (req, res, ctx) => {
        console.log('MOCK error /email');
        const noData = {
            "error": {
                "message": "no data",
                "status": 400
            }
        }
        return res(ctx.json(noData));
    }),

]