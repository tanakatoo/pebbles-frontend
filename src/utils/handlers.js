import { rest } from 'msw'

let loginResponse = {
    token: 'abcdefg',
    user: {
        id: 2,
        username: 'hello',
        role: 'regular',
        premium_join_date: '',
        premium_end_date: '',
        free_trial_start_date: '',
    }
}

const contactsResponse =
    [
        {
            "user_id": 1,
            "username": "ktoo",
            "avatar": "2.jpg"
        },
        {
            "user_id": 15,
            "username": "june",
            "avatar": "4.jpg"
        }
    ]

const contactsResponseAfterBlocking =
    [
        {
            "user_id": 1,
            "username": "ktoo",
            "avatar": "2.jpg"
        }
    ]

const blockedResponse = [
    {
        "blocked_user_id": 3,
        "username": "blockMe",
        "avatar": "9.jpg"
    }
]

const unblockedResponse = [

]

const mainSearchResults = [
    {
        "username": "hello2",
        "avatar": "9.jpg",
        "name": null,
        "about": null,
        "premium_acct_id": null,
        "end_date": null
    },
    {
        "username": "helloo",
        "avatar": "14.jpg",
        "name": null,
        "about": null,
        "premium_acct_id": null,
        "end_date": null
    }
]

const changePassword = "completed"

const locations = [{ description: "abcd", place_id: "1234" }]

const newMessage = [
    {
        "toavatar": "9.jpg",
        "to": "hello",
        "fromavatar": "",
        "from": "hello"
    }
]

const continuingMsg = [
    {
        "toavatar": "9.jpg",
        "fromavatar": "2.jpg",
        "from": "ktoo",
        "to": "hello",
        "from_user_id": 1,
        "to_user_id": 2,
        "msg": "the most recent one reply to second one",
        "sent_at": "2023-06-09 15:30:00",
        "read": true
    },
    {
        "toavatar": "2.jpg",
        "fromavatar": "9.jpg",
        "from": "hello",
        "to": "ktoo",
        "from_user_id": 2,
        "to_user_id": 1,
        "msg": "2nd most recent one reply",
        "sent_at": "2023-06-08 15:30:00",
        "read": true
    },
    {
        "toavatar": "9.jpg",
        "fromavatar": "2.jpg",
        "from": "ktoo",
        "to": "hello",
        "from_user_id": 1,
        "to_user_id": 2,
        "msg": "first message",
        "sent_at": "2023-06-07 15:30:00",
        "read": true
    }
]

const latestMessages = [
    {
        "toavatar": "9.jpg",
        "fromavatar": "4.jpg",
        "from": "june",
        "to": "hello",
        "from_user_id": 15,
        "to_user_id": 2,
        "msg": "testing\n",
        "sent_at": "2023-07-15 20:57:44.804634",
        "read": false
    },
    {
        "toavatar": "9.jpg",
        "fromavatar": "2.jpg",
        "from": "ktoo",
        "to": "hello",
        "from_user_id": 1,
        "to_user_id": 2,
        "msg": "the most recent one reply to second one",
        "sent_at": "2023-06-09 15:30:00",
        "read": true
    }
]

const returnUser = [
    {
        "id": 2,
        "username": "hello",
        "name": null,
        "avatar": "9.jpg",
        "country_en": null,
        "country_ja": null,
        "city_en": null,
        "city_ja": null,
        "state_en": null,
        "state_ja": null,
        "gender": "female",
        "about": null,
        "free_trial_start_date": null,
        "study_buddy_bio": "just chatting",
        "study_buddy_purpose": "i don't know why",
        "native_language": "Japanese",
        "learning_language": "English",
        "language_level": "Beginner",
        "time_zone": "NewYork",
        "age_range": "26-35",
        "study_buddy_active": true,
        "study_buddy_types": ['Volunteer'],
        "study_buddy_activate_date": "2023-07-13T04:00:00.000Z"
    }
]

const myProfile = {
    "id": 2,
    "username": "hello",
    "name": "",
    "email": "karmen.tanakaa@gmail.com",
    "role": "regular",
    "avatar": "9.jpg",
    "sign_up_date": "2023-05-01T00:00:00.000Z",
    "last_login_date": "2023-05-01T00:00:00.000Z",
    "language_preference": "English",
    "country_en": "",
    "country_ja": "",
    "city_en": "",
    "city_ja": "",
    "state_en": "",
    "state_ja": "",
    "gender": "male",
    "about": "",
    "myway_advice": "",
    "myway_habits": "",
    "motivational_level": "",
    "study_time": "",
    "premium_join_date": null,
    "premium_end_date": null,
    "raz_reading_level": "",
    "study_buddy_bio": "",
    "native_language": "",
    "learning_language": "",
    "language_level": "",
    "myway_language_level": "",
    "time_zone": "",
    "age_range": "",
    "study_buddy_active": false,
    "study_buddy_purpose": "",
    "free_trial_start_date": null,
    "study_buddy_activate_date": null,
    "study_buddy_types": [],
    "goals": [],
    "myProfile": true
}


export const handlers = [

    //login
    rest.post('http://localhost:3001/auth/login', (req, res, ctx) => {

        console.log(res(ctx.json(loginResponse)))
        return res(ctx.json(loginResponse));
    }),

    rest.post('http://localhost:3001/auth/logout', (req, res, ctx) => {

        console.log('MOCK /logout')
        window.localStorage.setItem('token', null)
        return res(ctx.json());
    }),


    // Called via <ExploreCommunity type="studyBuddy" />
    rest.get('http://localhost:3001/study-buddies/1', (req, res, ctx) => {
        console.log('MOCK /study-buddies/1');

        return res(ctx.json([]));
    }),


    rest.post('http://localhost:3001/auth/register', (req, res, ctx) => {
        console.log('MOCK /auth/register');

        return res(ctx.json(loginResponse));
    }),


    rest.post('http://localhost:3001/email', (req, res, ctx) => {
        console.log('MOCK /email');

        return res(ctx.json('sents'));
    }),

    //for blocking users
    rest.get('http://localhost:3001/users/contacts', (req, res, ctx) => {
        console.log('MOCK /users/contacts');

        return res(ctx.json(contactsResponse));
    }),

    //for unblocking users
    rest.get('http://localhost:3001/users/blocked', (req, res, ctx) => {
        console.log('MOCK /users/blocked');

        return res(ctx.json(blockedResponse));
    }),


    rest.get('http://localhost:3001/users/block', (req, res, ctx) => {
        console.log('MOCK /users/block');

        return res(ctx.json(blockedResponse));
    }),

    //block a user
    rest.post('http://localhost:3001/users/block/:username', (req, res, ctx) => {
        console.log('MOCK /users/block/:username');

        return res(ctx.json(contactsResponseAfterBlocking));
    }),

    rest.delete('http://localhost:3001/users/unblock/:username', (req, res, ctx) => {
        console.log('MOCK delete /users/unblock');

        return res(ctx.json(unblockedResponse));
    }),



    rest.get('http://localhost:3001/users/search', (req, res, ctx) => {
        console.log('MOCK /users/search');

        return res(ctx.json(mainSearchResults));
    }),

    rest.post('http://localhost:3001/auth/change-password', (req, res, ctx) => {
        console.log('MOCK /auth/change-password');

        return res(ctx.json(changePassword));
    }),

    //for continuing conversation
    rest.get('http://localhost:3001/messages/:username', (req, res, ctx) => {
        console.log('MOCK /messages/:username');
        console.log(req.params)
        return res(ctx.json(continuingMsg));
    }),


    //for continuing conversation
    rest.post('http://localhost:3001/messages/:username/send', (req, res, ctx) => {
        console.log('MOCK /messages/:username/send');
        console.log(req.params)
        return res(ctx.json(continuingMsg));
    }),
    //get latest message from all users
    rest.get('http://localhost:3001/messages', (req, res, ctx) => {
        console.log('MOCK /messages');
        return res(ctx.json(latestMessages));
    }),

    //get latest message from all users
    rest.get('http://localhost:3001/users/saved-users', (req, res, ctx) => {
        console.log('MOCK /users/saved-users');
        return res(ctx.json(returnUser));
    }),


    rest.post('http://localhost:3001/auth/set-password', (req, res, ctx) => {
        console.log('MOCK /auth/set-password');
        return res(ctx.json(loginResponse));
    }),

    rest.get('http://localhost:3001/users/hello', (req, res, ctx) => {
        console.log('MOCK /users/hello');
        return res(ctx.json(myProfile));
    }),

    //mock the google api
    rest.get('http://localhost:3001/external/api/location', (req, res, ctx) => {
        console.log('MOCK /external/api/location');
        return res(ctx.json(locations));
    }),

    //update data
    rest.patch('http://localhost:3001/users/:username', (req, res, ctx) => {
        console.log('MOCK patch /users/:username');
        return res(ctx.json(myProfile));
    })
]