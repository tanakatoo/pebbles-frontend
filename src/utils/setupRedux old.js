
import { configureStore } from '@reduxjs/toolkit';
import rootReducer from "../reducers/rootReducer"
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';

import storage from 'redux-persist/lib/storage'
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2'

export const persistConfig = {
    key: 'root', // Key prefix for the persisted state
    storage, // Storage mechanism (e.g., local storage)
    stateReconciler: autoMergeLevel2, // Specify which parts of the state to persist
    whitelist: ['profile', 'langFont']
};

export const persistedReducer = persistReducer(persistConfig, rootReducer);
export const store = configureStore({
    reducer: persistedReducer,
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware({
            serializableCheck: {
                ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER]
            }
        })
})
export const persistedStore = persistStore(store)