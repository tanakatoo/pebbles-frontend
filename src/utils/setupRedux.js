import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
// As a basic setup, import your same slice reducers
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import { PersistGate } from 'redux-persist/integration/react'
import { ListSlice } from "../reducers/rootReducer";
import rootReducer from "../reducers/rootReducer"
import storage from 'redux-persist/lib/storage'
import storeSetup from "./storeSetup"
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2'

// const persistConfig = {
//     key: 'root', // Key prefix for the persisted state
//     storage, // Storage mechanism (e.g., local storage)
//     stateReconciler: autoMergeLevel2, // Specify which parts of the state to persist
//     whitelist: ['profile', 'langFont']
// };
// const persistedReducer = persistReducer(persistConfig, rootReducer);

// const persistedStore = persistStore(store)
// const root = ReactDOM.createRoot(document.getElementById('root'));


export function renderWithProviders(
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = storeSetup(preloadedState),
        ...renderOptions
    } = {}
) {
    function Wrapper({ children }) {

        return <Provider store={store}>{children}</Provider>;
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}