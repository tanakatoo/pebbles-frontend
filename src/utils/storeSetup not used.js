import { combineReducers, configureStore } from '@reduxjs/toolkit'

import rootReducer from '../reducers/rootReducer'
// import userReducer from '../features/users/userSlice'

// Create the root reducer separately so we can extract the RootState type
// const rootReducer = combineReducers({
//     user: userReducer
// })

export const setupStore = preloadedState => {
    return configureStore({
        reducer: rootReducer,
        preloadedState
    })
}