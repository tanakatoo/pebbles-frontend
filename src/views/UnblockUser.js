import React, { useEffect, useState } from 'react'
import UserApi from '../api/user'
import ServerError from '../components/form/ServerError'
import Spinner from '../components/common/Spinner'
import { v4 as uuid } from "uuid"
import useSetToken from '../hooks/useSetToken'
import SearchBar from '../components/common/SearchBar'
import usePageText from "../hooks/usePageText"
import Protected from '../components/common/Protected'
import { useSelector } from 'react-redux'
import AvatarWithName from '../components/common/AvatarWithName'
import StickyBar from '../components/common/StickyBar'
import { useNavigate } from 'react-router-dom'
import { Formik, Form, Field, ErrorMessage } from 'formik'
import NoData from '../components/common/NoData'
import blockUnblockSchema from '../components/form/validation/blockUnblock'
import TextError from '../components/form/errorComponents/TextError'

function UnblockUser() {
    const navigate = useNavigate()
    const [data, setData] = useState([])
    const [errors, setErrors] = useState([])
    const [token = null] = useSetToken()
    const [doneGettingData, setDoneGettingData] = useState(false)
    const [pageText, lang] = usePageText("messages")


    const getData = async () => {
        try {
            setDoneGettingData(false)
            const res = await UserApi.getBlocked()
            setData(res)
            setDoneGettingData(true)
        } catch (e) {
            if (e instanceof TypeError) {

                //means server is down
                setErrors(["UNKNOWN"])
            } else {

                setErrors(e)
            }
        }
    }

    useEffect(() => {

        getData()
    }, [])

    const handleSearch = () => {
        //to be implemented later
        console.log('call to search for contact')
    }

    const handleCancel = () => {
        navigate('/messages')
    }

    const INITIAL_DATA = { username: '' }
    let options = []
    if (data.length > 0) {
        data.map(d => options.push({ key: d.username, value: d.username }))

    }

    return (
        <Protected>
            <div className='mb-12 flex justify-center '>

                {errors.length > 0 && <ServerError msg={errors} />}
                {data.length === 0 && errors.length === 0 && doneGettingData === false ?
                    <Spinner />
                    : doneGettingData === true && data.length === 0 ?
                        <NoData msg={pageText.NO_CONTACTS_UNBLOCK} link='/messages' linkText={pageText.BACK_TO_CONTACT_LINK} />

                        :
                        errors.length === 0 ?
                            <div className='w-full grid grid-cols-4 gap-x-4 pb-6 md:grid-cols-8 lg:grid-cols-12  
                        lg:max-w-[900px] content-center shadow-md '>
                                {/* <div className='col-span-full mb-4 mt-2'><SearchBar handleSearch={handleSearch} /></div> */}
                                <Formik
                                    initialValues={INITIAL_DATA}
                                    validationSchema={blockUnblockSchema}
                                    onSubmit={async (values, onSubmitProps) => {
                                        setErrors([])
                                        try {
                                            console.log('vlues to submit', values)
                                            const res = await UserApi.unblockUser(values.username)
                                            getData()
                                        } catch (e) {
                                            if (e instanceof TypeError) {
                                                //means server is down
                                                console.log(e)
                                                setErrors(["UNKNOWN"])
                                            } else {
                                                console.log(e)
                                                setErrors(e)
                                            }
                                        } finally {
                                            onSubmitProps.setSubmitting(false)
                                            onSubmitProps.resetForm()
                                        }
                                    }}
                                >{formik => {
                                    return (
                                        <>
                                            <div className='col-span-full w-full m-4 flex gap-4'>
                                                <ErrorMessage name='username' component={TextError} exclamation={true} />
                                            </div>
                                            <div className='col-span-full sticky top-0 mb-4'>
                                                <StickyBar action={pageText.UNBLOCK} backText={pageText.BACK} onClickX={handleCancel} onClick={formik.submitForm} />
                                            </div>
                                            <div className='col-span-full '>
                                                <Form className='w-full justify-end'>
                                                    {data.map(d => {
                                                        return (
                                                            <React.Fragment key={uuid()}>
                                                                <div className='mx-4 py-3 flex items-center '>
                                                                    <div className='grow'>
                                                                        <AvatarWithName src={`../avatars/${d.avatar}`} username={d.username} />
                                                                    </div>

                                                                    <Field name='username' type="radio" value={d.username}>{
                                                                        ({ field }) => < input type="radio" {...field} />
                                                                    }
                                                                    </Field>

                                                                </div >
                                                            </React.Fragment>
                                                        )
                                                    })}
                                                </Form>
                                            </div>


                                        </>
                                    )
                                }
                                    }
                                </Formik >
                            </div >
                            : ''
                }

            </div >
        </Protected >
    )
}

export default UnblockUser