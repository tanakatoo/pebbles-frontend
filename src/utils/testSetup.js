// import { persistedStore, store } from './setupRedux';
import { render } from '@testing-library/react'
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react'
import { setupTestStore } from './setupRedux';


const renderWithProviders = (ui, options = {}) => {
    const [store, persistedStore] = setupTestStore(options.preloadedState);

    const AllProviders = ({ children }) => {
        return (
            <Provider store={store}>
                <PersistGate loading={null} persistor={persistedStore}>
                    {children}
                </PersistGate>
            </Provider>
        )
    };

    return { store, ...render(ui, { wrapper: AllProviders, ...options }) };
}


// "re-export everything"
export * from '@testing-library/react';

export { renderWithProviders };