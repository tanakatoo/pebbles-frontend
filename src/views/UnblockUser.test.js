import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { renderWithProviders } from '../utils/testSetup';
import UnblockUser from './UnblockUser';
import App from '../App';
import { waitFor, act, screen } from '../utils/testSetup';
import { fireEvent } from '../utils/testSetup';
import { errorHandlers } from '../utils/errorHandlers';
import { server } from '../utils/server'
import { rest } from 'msw';

// Mock window.scrollTo()
window.scrollTo = jest.fn();

afterAll(() => {
    jest.clearAllMocks();
});


beforeEach(async () => {
    window.localStorage.clear()

})

afterEach(() => {
    window.localStorage.clear()
})

let loginResponse = {
    token: 'abcdefg',
    profile: {
        id: 2,
        username: 'hello',
        name: 'Hello World',
        email: "karmen.tanakaa@gmail.com",
        role: 'regular',
        premium_join_date: '',
        premium_end_date: '',
        free_trial_start_date: '',
        study_buddy_types: [],
        study_buddy_active: false
    }
};


describe('UnblockUser', () => {
    test('renders without crashing', () => {
        renderWithProviders(
            <MemoryRouter initialEntries={["/users/unblock"]}>
                <App />
            </MemoryRouter>,
            {
                preloadedState: { profile: loginResponse }
            }
        );
    });

    test('matches snapshot', async () => {
        const { asFragment } = renderWithProviders(
            <MemoryRouter initialEntries={["/users/unblock"]}>
                <App />
            </MemoryRouter>,
            {
                preloadedState: { profile: loginResponse }
            }
        );

        expect(asFragment()).toMatchSnapshot();
    });


    test('able to Unblock a user', async () => {

        const { findByRole, findByText } =

            renderWithProviders(
                <MemoryRouter initialEntries={["/users/unblock"]}>
                    <App />
                </MemoryRouter>,
                {
                    preloadedState: { profile: loginResponse }
                }
            );


        let myRadio = await findByRole('radio', { value: { text: 'blockMe' } })
        fireEvent.click(myRadio)

        let btn = await findByText(/Unblock/)
        fireEvent.click(btn);

        //this one doesn't work
        // server.use(
        //     rest.get('http://localhost:3001/users/blocked', (req, res, ctx) => {
        //         console.log('MOCK no data /users/blocked');

        //         return res.once(ctx.json([]));
        //     }),
        // )

        // myRadio = await findByText(/blockMe/)
        // expect(myRadio).not.toBeInTheDocument();

    })
})
