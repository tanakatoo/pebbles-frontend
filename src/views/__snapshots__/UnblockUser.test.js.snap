// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnblockUser matches snapshot 1`] = `<DocumentFragment />`;
